proto/go_proto_library.bzl

```diff
@@ -0,0 +1,238 @@
+          ] + [
+              f.path for f in ctx.files.src
+          ],
+      progress_message="Generating into %s" % ctx.outputs.out.short_path,
+      mnemonic="GoProtocGen",
+      env = {"PATH": ctx.files.protoc_gen_go[0].dirname},
+      executable=ctx.executable.protoc)
+  # This is the current hack for files without 'option go_package'
+  # Generate into .pb.go, then cp into "real location"
+  if bazel_style:
+    ctx.action(
+        inputs=[proto_out],
+      outputs=[ctx.outputs.out],
+        command="cp %s %s" % (proto_out.path, ctx.outputs.out.path),
+        mnemonic="GoProtocGenCp")
+  return struct(_protos=protos,
+                _m_import_path=m_import_path,
+                name=(ctx.label.package + go_package_name))
+
+go_proto_library_gen = rule(
+    attrs = {
+        "deps": attr.label_list(),
+        "src": attr.label(
+            mandatory = True,
+            single_file = True,
+            allow_files = True,
+        ),
+        "grpc": attr.int(default = 0),
+        "out": attr.output(mandatory = True),
+        "protoc": attr.label(
+            default = Label("@com_google_protobuf//:protoc"),
+            executable = True,
+            single_file = True,
+            allow_files = True,
+            cfg = "host",
+        ),
+        "protoc_gen_go": attr.label(
+            default = Label("@com_github_golang_protobuf//protoc-gen-go"),
+            allow_files = True,
+            cfg = "host",
+        ),
+        "_protos": attr.label_list(default = []),
+        "go_prefix": attr.label(
+            providers = ["go_prefix"],
+            default = Label(
+                "//:go_prefix",
+                relative_to_caller_repository = True,
+            ),
+            allow_files = False,
+            cfg = "host",
+        ),
+    },
+    implementation = _go_proto_library_gen_impl,
+)
+
+def _add_target_suffix(target, suffix):
+  idx = target.find(":")
+  if idx != -1:
+    return target + suffix
+  toks = target.split("/")
+  return target + ":" + toks[-1] + suffix
+
+def go_proto_library(name = None, srcs = None, deps = None,
+                     has_services = 0,
+                     testonly = 0, visibility = None,
+                     **kwargs):
+  """Macro which generates and compiles protobufs for Go.
+
+  Args:
+    name: name assigned to the underlying go_library,
+          typically "foo_proto" for ["foo.proto"]
+    srcs: a list of .proto source files, currently only 1 supported
+    deps: a mixed list of either go_proto_libraries, or
+          any go_library which has a companion
+          filegroup(name=name+"_protos",...)
+          which contains the protos which were used
+    has_services: indicates the proto has gRPC services and deps
+    testonly: mark as testonly
+    visibility: visibility to use on underlying go_library
+  """
+  if not name:
+    fail("name is required")
+  if not srcs or len(srcs) != 1:
+    fail("exactly 1 src is required")
```

Comment:
Use the second parameter of `fail` to report the source attribute name of the error.